gdx/src/com/badlogic/gdx/graphics/g2d/TextureArraySpriteBatch.java

```diff
@@ -0,0 +1,1414 @@
+
+	/** Number of rendering calls, ever. Will not be reset unless set manually. **/
+	public int totalRenderCalls = 0;
+
+	/** The maximum number of sprites rendered in one batch so far. **/
+	public int maxSpritesInBatch = 0;
+
+	/** The current number of textures in the LFU cache. Gets reset when calling {@link#begin()} **/
+	private int currentTextureLFUSize = 0;
+
+	/** The current number of texture swaps in the LFU cache. Gets reset when calling {@link#begin()} **/
+	private int currentTextureLFUSwaps = 0;
+
+	/** Constructs a new TextureArraySpriteBatch with a size of 1000, one buffer, and the default shader.
+	 * @see TextureArraySpriteBatch#TextureArraySpriteBatch(int, ShaderProgram) */
+	public TextureArraySpriteBatch () {
+		this(1000);
+	}
+
+	/** Constructs a TextureArraySpriteBatch with one buffer and the default shader.
+	 * @see TextureArraySpriteBatch#TextureArraySpriteBatch(int, ShaderProgram) */
+	public TextureArraySpriteBatch (int size) {
+		this(size, null);
+	}
+
+	/** Constructs a new TextureArraySpriteBatch. Sets the projection matrix to an orthographic projection with y-axis point
+	 * upwards, x-axis point to the right and the origin being in the bottom left corner of the screen. The projection will be
+	 * pixel perfect with respect to the current screen resolution.
+	 * <p>
+	 * The defaultShader specifies the shader to use. Note that the names for uniforms for this default shader are different than
+	 * the ones expect for shaders set with {@link #setShader(ShaderProgram)}. See {@link#createDefaultShader()}.
+	 * @param size The max number of sprites in a single batch. Max of 8191.
+	 * @param defaultShader The default shader to use. This is not owned by the TextureArraySpriteBatch and must be disposed
+	 *           separately. */
+	public TextureArraySpriteBatch (int size, ShaderProgram defaultShader) {
+
+		// 32767 is max vertex index, so 32767 / 4 vertices per sprite = 8191 sprites max.
+		if (size > 8191) throw new IllegalArgumentException("Can't have more than 8191 sprites per batch: " + size);
+
+		// Query the number of available texture units and decide on a safe number of texture units to use
+		IntBuffer texUnitsQueryBuffer = BufferUtils.newIntBuffer(32);
+
+		Gdx.gl.glGetIntegerv(GL20.GL_MAX_TEXTURE_IMAGE_UNITS, texUnitsQueryBuffer);
+
+		maxTextureUnits = texUnitsQueryBuffer.get();
+
+		// Some OpenGL drivers (I'm looking at you, Intel!) do not report the right values,
+		// so we take caution and test it first, reducing the number of slots if needed.
+		if (defaultShader == null) {
+
+			// Will try to create a shader with a lower amount of texture units if creation fails.
+			while (maxTextureUnits > 0) {
+
+				try {
+
+					shader = createDefaultShader(maxTextureUnits);
+
+					break;
+
+				} catch (Exception e) {
+					maxTextureUnits /= 2;
+				}
+			}
+
+			if (maxTextureUnits == 0) {
+				throw new IllegalStateException("Texture Arrays are not supported on this device.");

```

Comment:
Is it possible to make some kind of automatic rollback to the good old SpriteBatch if this one isn't compatible?